x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -349,46 +349,23 @@ public void authorizeIndexAction(
-                // all wildcard expressions have been resolved and only the security plugin could have set '-*' here.
-                // '-*' matches no indices so we allow the request to go through, which will yield an empty response
-                if (resolvedIndices.isNoIndicesPlaceholder()) {
-                    // check action name
-                    listener.onResponse(authorizeIndexActionName(action, authorizationInfo, IndicesAccessControl.ALLOW_NO_INDICES));
-                } else {
-                    assert resolvedIndices.getLocal().stream().noneMatch(Regex::isSimpleMatchPattern)
-                        || replaceable.indicesOptions().expandWildcardExpressions() == false
-                        || (request instanceof AliasesRequest aliasesRequest && aliasesRequest.expandAliasesWildcards() == false)
-                        : "expanded wildcards for local indices OR the request should not expand wildcards at all";
-                    listener.onResponse(
-                        buildIndicesAccessControl(
-                            action,
-                            authorizationInfo,
-                            Sets.newHashSet(resolvedIndices.getLocal()),
-                            aliasOrIndexLookup
-                        )
-                    );
-                }
-            }, listener::onFailure));
         } else {
             try {
-                final IndexAuthorizationResult indexAuthorizationResult = authorizeIndexActionName(
-                    action,
-                    authorizationInfo,
-                    IndicesAccessControl.ALLOW_NO_INDICES
-                );
-                if (indexAuthorizationResult.isGranted()) {
+                if (request instanceof IndicesRequest.Replaceable replaceable && replaceable.allowsRemoteIndices()
+                    || ensureRBAC(authorizationInfo).getRole().checkIndicesAction(action)) {
                     indicesAsyncSupplier.getAsync(ActionListener.wrap(resolvedIndices -> {
                         assert resolvedIndices.isEmpty() == false
                             : "every indices request needs to have its indices set thus the resolved indices must not be empty";
                         // all wildcard expressions have been resolved and only the security plugin could have set '-*' here.
                         // '-*' matches no indices so we allow the request to go through, which will yield an empty response
                         if (resolvedIndices.isNoIndicesPlaceholder()) {
-                            listener.onResponse(new IndexAuthorizationResult(true, IndicesAccessControl.ALLOW_NO_INDICES));
+                            if (request instanceof IndicesRequest.Replaceable replaceable
+                                && replaceable.allowsRemoteIndices()
+                                && ensureRBAC(authorizationInfo).getRole().checkIndicesAction(action) == false) {
```

Comment:
I had some hard time to understand this `if` check because it has 3 conditions and is nested under an outer `if` check which uses the same conditions but different logical operations. I then realised it is functionally equivalent to this simpler version:

```java
if (ensureRBAC(authorizationInfo).getRole().checkIndicesAction(action) == false) {
```

However, the simpler version has a performance implication in that the `checkIndicesAction` can sometimes execute twice (1st time in the outer `if` and 2nd time here). So it is not really viable. But I think it would have helped me to understasnd it quicker if the conditions are wrapped and arranged a bit differently, e.g. something like:

```java
final boolean allowsRemoteIndices = request instanceof IndicesRequest.Replaceable replaceable && replaceable.allowsRemoteIndices();
if (allowsRemoteIndices || isActionGranted(authorizationInfo, action)) {
  indicesAsyncSupplier.getAsync(ActionListener.wrap(resolvedIndices -> {
    if (resolvedIndices.isNoIndicesPlaceholder()) {
      if (allowsRemoteIndices) {
        listener.onResponse(new IndexAuthorizationResult(true, isActionGranted(authorizationInfo, action) ? IndicesAccessControl.ALLOW_NO_INDICES : IndicesAccessControl.DENIED));
      } else {
        // Paranoid assertion to make sure the outer `if` check always holds in case of future refactoring
        assert isActionGranted(authorizationInfo, action);
        listener.onResponse(new IndexAuthorizationResult(true, IndicesAccessControl.ALLOW_NO_INDICES));
      }
    } else {
       ...
    }
  }
}
...

private boolean isActionGranted(AuthorizationInfo authorizationInfo, String action) {
    return ensureRBAC(authorizationInfo).getRole().checkIndicesAction(action);
}
```

The above is more readable to me personally. We could also have constants for `IndexAuthorizationResult` when it is either `DENIED` or `ALLOW_NO_INDICES` which can help tidy up the code a bit more.

Reply:
Yes, I think it makes sense to group the conditions as you describe. And double Yes for 
> We could also have constants for IndexAuthorizationResult when it is either DENIED or ALLOW_NO_INDICES which can help tidy up the code a bit more.

I can bake them both in this PR if you agree.

Reply:
+1 please go ahead